### ISSUE TITLE
508-defect-2 [SCREENREADER, COGNITION]: Table labels SHOULD be concise and descriptions SHOULD use aria-describedby

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#17149]
# [508-defect-2](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-2)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

<hr/>

## Point of Contact

**VFS Point of Contact:** Josh

## User Story or Problem Statement

As a screen reader user, I expect to hear concise labels for tables so that I can understand them efficiently.

## Details

As of now, tables are labelled by two `aria-labelledby` ids:
- The heading of the section 👍 
- The description of the section 👎 

Because the description of the section is added to the label, the table's name becomes too verbose for screen reader users.

<img width="300" alt="Screen Shot 2020-12-09 at 9 32 34 AM" src="https://user-images.githubusercontent.com/14154792/101642849-7a875580-3a01-11eb-97ff-0aae9863ca6e.png">

This can easily be fixed by appropriately mapping the description of the section to `aria-describedby` instead of appending it to the label.

Furthermore, the table does not describe to the screen reader user the string "Displaying X - Y of Z" which is valuable information. Sighted users will be able to notice this change every time they move through the pagination or if they were to jump directly to the table itself using a screen reader shortcut. In order to provide this to our screen reader users we should:
- Add an `aria-live="polite"` to this `p`
- Add a unique id to this `p` and use it to describe the table using `aria-describedby`

Edit: A new issue I recently caught was misspelled IDs between "received" and "recieved", which caused an issue to be caught by axe 
<img width="849" alt="Screen Shot 2022-12-16 at 1 49 39 PM" src="https://user-images.githubusercontent.com/14154792/208168299-13bac2d8-4d98-4658-ab9e-71f630128712.png">

## Tasks
- [x] Provide recommendations to platform on how this can be implemented

## Acceptance Criteria
- [x] Platform has been notified of this 
- [x] [Platform has added `aria-describedby` prop](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/378#issuecomment-1097076620)
- [ ] Both table's `aria-labelledby` matches the `id` of the section's heading.
- [ ] Both table's `aria-describedby` matches the description below the section's heading
- [ ] Platform reaches a conclusion on [how to announce changes for pagination based on this bug ticket](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/1408)
- [ ] Both table's "displaying x out of y" are announced to the user when the user navigates to a different table page using the pagination

## Environment

* Operating System: Mac, Windows
* Browser: Safari, Firefox
* Screenreading device: VoiceOver, NVDA
* Server destination: staging, production

## Steps to Recreate

1. [Log into user 54's account and open the payment history page](https://staging.va.gov/va-payment-history/payments/?postLogin=true).
2. Inspect the table using developer tools.
3. Confirm it has an `aria-labelledby` that points to both the heading and description of the section.

## Solution

There are two tables on this page, so don't forget to update both of them 😄 

### Before
```html
<!-- The paragraph that describes how many items exist does not semantically describe the table (yet) -->
<p class="vads-u-font-size--lg vads-u-font-family--serif">Displaying 55 - 60 of 67</p>

<!-- The table includes its description under aria-labelled which we'll have to change -->
<table 
  aria-labelledby="paymentsReceivedHeader paymentsReceivedContent" 
  class="responsive" 
  role="table">
```

### After (aria-live)
``` diff
<!-- Add a unique id to this paragraph that describes how many items are displaying -->
<!-- I've also added "payments" to this string so it's clear to the user what is being displayed -->
<!-- To ensure this gets announced when the user navigates the pagination, I've also added aria-live-->
+ <p id="unique-id" aria-live="polite" class="vads-u-font-size--lg vads-u-font-family--serif">
+     Displaying 55 - 60 of 67 payments.
+ </p>

<!-- Add the unique id into an aria-describedby and move the descriptive id from aria-labelledby into aria-describedby.-->
<table 
  aria-labelledby="paymentsReceivedHeader" 
+  aria-describedby="unique-id paymentsReceivedContent" 
  class="responsive" 
  role="table">
```

### After (focus management)
We can also opt to use focus management instead of an aria-polite to announce changes, but this may unexpectedly move the user away from what they are doing. 

``` diff
<!-- Add a unique id to this paragraph that describes how many items are displaying -->
<!-- I've also added "payments" to this string so it's clear to the user what is being displayed -->
<!-- To ensure this gets announced when the user navigates the pagination, I've also added aria-live-->
+ <p id="unique-id" class="vads-u-font-size--lg vads-u-font-family--serif">
+     Displaying 55 - 60 of 67 payments.
+ </p>

<!-- Add the unique id into an aria-describedby and move the descriptive id from aria-labelledby into aria-describedby.-->
<table 
  aria-labelledby="paymentsReceivedHeader" 
+  aria-describedby="unique-id paymentsReceivedContent" 
  class="responsive" 
  role="table">
```

Some notes here:
- I've placed the "Displaying 55-60..." string ahead of the longer table description so it does not get buried in verbose content.
- While testing, I noticed that adding an `id` and `aria-live` to the `p` in question causes it to stop updating. I'm unsure as to the reason why and could use some help debugging there.

## WCAG or Vendor Guidance

[WAI-ARIA 1.1: The aria-labelledby attribute is similar to aria-describedby in that both reference other elements to calculate a text alternative, but a label should be concise, where a description is intended to provide more verbose information.](https://www.w3.org/TR/wai-aria/#aria-labelledby)

**Comment by jason-gcio** [comment]
Bringing this in to Grooming for more discussion

**Comment by jason-gcio** [comment]
@joshkimux - we are going to provide some recommendations on implementing this at the platform level and are keeping you on as assignee for now for awareness

**Comment by joshkimux** [comment]
@jason-gcio Awesome that's perfect!

**Comment by micahchiang** [comment]
Feature request made to vsp design system team here - https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/378

**Comment by micahchiang** [comment]
Moving this to validate for the time being, since the purpose of this ticket was to make recommendations to the platform team.

**Comment by micahchiang** [comment]
Closing this, as the ticket mentioned in the comment above has been assigned to the right people on the vsp-design-system-team

**Comment by joshkimux** [comment]
Re-opening this, as it has not yet been resolved
